Unsharing pattern types

trefis opened the pull request:
Extracts the least controversial part #1675 .

trefis committed:
take an instance from type_pat, not from its caller

trefis committed:
add one test

trefis committed:
Changes

garrigue commented on typing/typecore.ml:
```diff
@@ -1085,13 +1085,18 @@ and type_pat_aux ~exception_allowed ~constrs ~labels ~no_existentials ~mode
   in
   let loc = sp.ppat_loc in
+  let rup k x =
+    if constrs = None then (ignore (rp x));
```
You record the type of the pattern before unification with the expected type.
This is a change w.r.t. the previous behavior.
Is this intentional?

garrigue replied:
Sorry, looking at what does `Stypes.record`, you are just registering the node earlier, and unification can still occur afterwards. So one has more information if the next unification fails, and the same otherwise. Fine.

garrigue commented on typing/typecore.ml:
```diff
@@ -1203,14 +1207,18 @@ and type_pat_aux ~exception_allowed ~constrs ~labels ~no_existentials ~mode
+      begin_def ();
+      let expected_ty = instance expected_ty in
+      end_def ();
+      generalize_structure expected_ty;
```
Here it should be OK to use `generalize`: since we are matching a value, different members could be seen as using independent instances.

trefis replied:
I don't think it makes a difference here: `expected_ty` doesn't contain any type variable at the generic level when we receive it. Which means that after the
```
begin_def;
instance;
end_def;
```
sequence, there won't be any generalizable variables.

Given that, I'd rather keep the call to `generalize_structure` here, for uniformity.

garrigue replied:
Oh, I see. I hadn't seen that due to the `generalize_structure` in `type_cases`, there are no generalized type variables in the type.
This would change if switch later to a more agressive policy of taking full instances, that could be fully generalized soundly.

But then it seems that the only improvement of the current patch is to avoid some warnings with label/constructor disambiguation. Is there any other practical change?

garrigue commented on typing/typecore.ml:
```diff
@@ -1271,15 +1279,21 @@ and type_pat_aux ~exception_allowed ~constrs ~labels ~no_existentials ~mode
+      generalize_structure expected_ty;
+      generalize_structure ty_res;
+      List.iter generalize_structure ty_args;
 
```
Here the sharing of type variables is needed, so this indeed has to be `generalize_structure`.

garrigue commented on typing/typecore.ml:
```diff
@@ -1334,27 +1352,34 @@ and type_pat_aux ~exception_allowed ~constrs ~labels ~no_existentials ~mode
         pat_loc = loc; pat_extra=[];
-        pat_type = expected_ty;
+        pat_type = record_ty;
         pat_attributes = sp.ppat_attributes;
```
Shouldn't it be `instance record_ty` ? And why call `rup` rather than `rp` (`record_ty` is already an instance of `expected_ty`) ?

trefis replied:
It should be `instance record_ty` you are correct.
However `record_ty` is not necessarily an instance of `expected_ty`, in fact in most cases it won't be: it will be only when `extract_concrete_record` succeeds.

trefis replied:
Note: I pushed a commit adding the instance taking.

garrigue reviewed (commented):
Globally, this looks correct. However, I'm wondering what should be returned in `pat_type`. We seem to have an invariant to keep it non-generic, but wouldn't it make sense to just always put the (generalized) `expected_ty`, from a semantical point of view?

trefis commented:
Thanks for the review. I haven't had time to think much about the question you ask (or to fix the small issues in the code) yet, but I'll get back to you some time next week.

garrigue commented:
I'm copying here an inline comment, as it might no be very visible:

Since, in `type_cases`, `generalize_structure` is called before calling `type_pattern`, `expected_ty` cannot contain generalized type variables.

As a result, it seems that the only improvement of the current patch is to avoid some warnings with label/constructor disambiguation. Is there any other practical change?

lpw25 commented:
I haven't actually tested it with this version of the patch, but I think it should address ambiguity errors like this one:

```ocaml
# type 'a r = { a : 'a; b: 'a };;
type 'a r = { a : 'a; b : 'a; }

# type 'a ty = Int : int ty | Float : float ty;;
type 'a ty = Int : int ty | Float : float ty

# let foo (type a) (ty : a ty) (x : a r) =
    match ty, x with
    | Int, { a = 3; b } -> b
    | _ -> assert false;;
```

trefis commented:
> I haven't actually tested it with this version of the patch, but I think it should address ambiguity errors like this one: [...]

Indeed, that's the test I add with my first commit: at that time it still complains that the type of `b` is ambiguous, but this is then fixed by the later commits.

> As a result, it seems that the only improvement of the current patch is to avoid some warnings with label/constructor disambiguation. Is there any other practical change?

The practical change is illustrated in the test suite by the example Leo just repeated.
As for the principality warnings regarding disambiguation: this PR doesn't contain changes to that effect. I wasn't sure whether we'd completely agreed on that point when we decided to split into smaller PRs. So I went with this small PR first, as it seemed to be the least controversial :)

I plan to open a new PR containing (only) the changes relating to principality warnings as a follow up to this PR.

trefis commented:
> However, I'm wondering what should be returned in `pat_type`. We seem to have an invariant to keep it non-generic, but wouldn't it make sense to just always put the (generalized) `expected_ty`, from a semantical point of view?

Perhaps. Although I think putting the instance makes just as much sense?
Also, in the implementation, in some cases we don't have a generic type at hand after typing the pattern, which I take as a hint that putting the non-generic type is the right thing to do.

garrigue commented:
My comment concerning whether `pat_type` should be generic is also for the future, if we propagate fully generic types: logically this type should be understood as the input of the pattern.

For now I'm satisfied with the current PR, which keeps all invariants, and so is safe.

trefis commented:
Understood.

Thanks for looking at this!

I've now rebased and squashed the commits, I'll merge once the CIs are happy (hopefully appveyor won't time out this time).
